contributing/Debugging.md

```diff
@@ -0,0 +1,168 @@
+# Debugging Gradle
+
+This guide is intended to get you started on how to debug the integration tests and different parts of the Gradle runtime in the Gradle code base.
+
+## TL:DR
+
+To debug a unit test, just run it in the debug mode.
+
+To debug an integration test, in most cases, all you need to do is to start listening debugger via the "Daemon debug" run configuration in IntelliJ IDEA and then run the test in debug mode with the `forkingIntegTest` task. 
+The debugger will automatically connect to the Gradle Daemon, and you can start debugging it.
+
+If you need to debug something besides the Daemon, read below.
+
+## Overview
+
+### IntelliJ IDEA setup
+
+For most cases, use [listening debugger](https://www.jetbrains.com/help/idea/attaching-to-local-process.html#attach-to-remote]).
+
+The build should create debugging run configurations for you [automatically](../build-logic/idea/src/main/kotlin/gradlebuild.ide.gradle.kts#L82), but you'll need to make sure to enable automatic restart of the debugging session once the debuggee is gone.
+Unfortunately, setting this up [is not automated yet](https://github.com/JetBrains/gradle-idea-ext-plugin/issues/84).
+
+![](./images/auto-restart-debugger.png)
+
+Before you start debugging, start the "Daemon debug" or "Debug Launcher" configuration. 
+It should be automatically re-used for subsequent test runs.
+
+Be careful when you stop the test: you should stop the test, not the debugger.
+
+### Gradle test executors
+
+There are [several executors](../testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/GradleContextualExecuter.java) available for running integration tests configured in Gradle.
+
+Most of them are forking, and you should be able to debug a test with any of them by running the test in the debug mode.
+The test detects that it is running in debug mode and provides all necessary information to the forked daemon to connect to the debugger.
+
+There is also an embedded executor that runs the test in the same JVM as the test runner.
+While it's convenient for debugging and sometimes may be faster, it's not recommended to use it because it may behave differently in some corner cases due to classloader issues.
```

Comment:
```suggestion
While this can be convenient for debugging and may offer faster execution, it's generally not recommended due to potential classloader issues, which may lead to different behavior in certain edge cases.
```